### PR TITLE
PreviewExecutor: Make sampling buffer bytes configurable

### PR DESCRIPTION
**Pull request by muga**
This PR enables users configuring bytes of sampling buffer that `preview` command tries to read from input data source. Since users have various files and they sometimes want to change the # of rows that `preview` shows, it's better to configure the bytes by users. In the latest version of Embulk, v0.8.18, the bytes is 32KB. For example, if an user has a TSV file that includes one record line and the record bytes is over 32KB, `preview` will fail by `NoSampleError`. 

To do that, it introduces `preview_sample_buffer_bytes` option. Users can configure sampling buffer bytes in Embulk config `exec:` section as following.

```
exec:
  preview_sample_buffer_bytes: 65536
in:
  type: file
  ... ...
```

### CODE REVIEW

**Comment by hiroyuki-sato** [comment]
Thank you for your great work!!.

IMO, A user may want to know current sample_size. 
So It is helpful to put it into the logger. 
If a user sees `32768 bytes` message, they want to change this value for a large file.
For example,

```
2017-03-15 21:11:55.609 +0900 [INFO] (0001:preview): Sampling the first 32768 bytes for preview...
```

Ref: An user request.
[#444 comment](https://github.com/embulk/embulk/issues/476#issuecomment-286402705)

**Comment by muga** [comment]
Thank you. I added the logging.

**Comment by hiroyuki-sato** [comment]
Thank you!. It must be helpful.

**Comment by dmikurube** [comment]
@muga What do you think about naming this more explicit? For example, `preview_sample_buffer_bytes`.

It can be confusing with existing `preview_sample_rows`. The word "size" may sound like the number of rows.

**Comment by muga** [comment]
@dmikurube Agree. I will rename with `preview_sample_buffer_bytes` that you mentioned. Once I finished writing tests, I will ask you to review this feature.

**Comment by hiroyuki-sato** [comment]
In short, It worked perfectly!!  👍 
Thanks!!.

By the way,  Can I use this configuration for `guess`?
(I would like to use this configuration in guess too.)

I think it doesn't need `preview_` prefix. `sample_buffer_bytes` is better. 

## Prepare test.

See detail (#471)
Test data
https://gist.github.com/kakoni/a1dfaafb1de76f82df938a9ba4ba400e


```
mkdir preview_test
cd preview_test
wget https://gist.githubusercontent.com/kakoni/a1dfaafb1de76f82df938a9ba4ba400e/raw/04d12b4862f6667403fde592b8016376e146e64a/config.yml
wget https://gist.githubusercontent.com/kakoni/a1dfaafb1de76f82df938a9ba4ba400e/raw/04d12b4862f6667403fde592b8016376e146e64a/fail.csv
```


## latest version (embulk 0.8.18)

```
2017-04-05 22:06:39.119 +0900 [INFO] (0001:preview): Listing local files at directory '.' filtering filename by prefix 'fail.csv'
2017-04-05 22:06:39.123 +0900 [INFO] (0001:preview): Loading files [fail.csv]
org.embulk.spi.DataException: Invalid record at line 39: 0000000,0000000,0000000,0000000,000,00,0,0,0,0,0,0,0,0,000000000000,00000000000000,0,0000,000000000000,00000000000000,0,0,0,0,0,00000000000,0000000000000000000000000000000000000000,0000,000000,00000000,00,00000,0000,0,00000,0000,0,00,0000000000,000000,0000,000000,0000,000000,0000,000000,0000,000000000000000000000000000000000000000000000000000000000000,0,00000,00000,0000000000,0000,0000000000,0,000000,0000,0000,0000,0000,000000,0000,00000,0000,0,00000,000000000000000000000000000000000000000000000000000,00,00,00,00,00000000000000000000000000000000000,00,000,0,000,000,000,000,000,000,000,00,00,0,00000000,00000000,00,000,0,00000,0,0000000,0000000,0,0,0,0,00000000000000,00000000000000,00000000000000,00000,0000,00000,0000,000,0000000,0000,0,0,0,0,0,0,0,0,000,0000,0000000,0000000,00000,0000000,00000000,00000,00000000,0000,0000000
	at org.embulk.standards.CsvParserPlugin.run(org/embulk/standards/CsvParserPlugin.java:368)
	at org.embulk.spi.FileInputRunner.run(org/embulk/spi/FileInputRunner.java:153)
	at org.embulk.exec.PreviewExecutor$2$1.run(org/embulk/exec/PreviewExecutor.java:122)
```


## preview_add_preview_sample_size_in_exec_config branch(44738ea54fd2fd4ae079244ef5b46bdfdc99e907)

Add entry

```yaml
exec:
  preview_sample_buffer_bytes: 1048576 # 1MB 1024 * 1024
```

```
/path/to/embulk/pkg/embulk-0.8.18.jar preview -G config.yml
2017-04-05 22:10:33.201 +0900: Embulk v0.8.18
2017-04-05 22:10:34.213 +0900 [INFO] (0001:preview): Listing local files at directory '.' filtering filename by prefix 'fail.csv'
2017-04-05 22:10:34.216 +0900 [INFO] (0001:preview): Loading files [fail.csv]
2017-04-05 22:10:34.228 +0900 [INFO] (0001:preview): Try to read 1,048,576 bytes from input source
*************************** 1 ***************************
     a (string) : a
     b (string) : b
     c (string) : c
     d (string) : d
     e (string) : e
     f (string) : f
     g (string) : g
     h (string) : h
```

**Comment by muga** [comment]
@hiroyuki-sato 

First, I'm now adding preview test framework on embulk-test. Because I don't want regression by this change. 

> Can I use this configuration for guess?
> (I would like to use this configuration in guess too.)

That's good idea 👍  I will create another issue ticket to do that.

> I think it doesn't need `preview_` prefix. `sample_buffer_bytes` is better.

I see. If we can use this option for `PreviewExecutor` and `GuessExecutor`, it would be even better I think. @dmikurube how about this?

**Comment by hiroyuki-sato** [comment]
@muga Thank you for your comment.

I just wanted to try this feature.
And I'm glad it worked as I expected.
I'll learn how to write test codes.

**Comment by dmikurube** [comment]
@muga In terms of naming `sample_buffer_bytes`, I basically agree with removing `preview_`, but we may need to consider consistency with existing `preview_sample_rows`?

**Comment by muga** [comment]
> but we may need to consider consistency with existing preview_sample_rows?

@dmikurube That's right. Thank you. I will fix it.

**Comment by muga** [comment]
Add tests by https://github.com/embulk/embulk/pull/601

**Comment by muga** [comment]
Rebased to use #601 for testing this feature.

**Comment by muga** [comment]
@dmikurube Please take a look this PR. I updated the PR's description as well.

I didn't remove `preview_` prefix from the option. We don't have a big benefit by sharing the option between `PreviewExecutor` and `GuessExecutor`. In the case of `preview_sample_rows` option, both of executors cannot share the option. because the meaning of "sample_rows" is different between both of executors.

**Comment by muga** [comment]
I'm still working on it.

**Comment by muga** [comment]
I noticed that `SamplingParserPlugin.runFileInputSampling` method uses `PreviewExecutorTask` and fixed it. The method is used by both of `preview` and `guess`.

File: embulk-core/src/main/java/org/embulk/exec/PreviewExecutor.java [hunk 1]
```diff
@@ -36,6 +36,10 @@
     public interface PreviewTask
             extends Task
     {
+        @Config("exec")
+        @ConfigDefault("{}")
+        public ConfigSource getExecConfig();
```
Comment (dmikurube) [inline embulk-core/src/main/java/org/embulk/exec/PreviewExecutor.java]: Any reason why we use `exec` for this while `preview_sample_rows` is out of `exec`?

Sorry, I was not very sure how `preview_sample_rows` is configured. Is it out of `exec`? In other words, the configurations is like following?

```
preview_sample_rows: 50
exec:
  preview_sample_buffer_bytes: 10000
in:
  ...
out:
  ...
```

Comment (muga) [inline embulk-core/src/main/java/org/embulk/exec/PreviewExecutor.java]: Sorry, I'm also not sure why the `PreviewTask` has `preview_sample_rows` option. We need to ask @frsyuki about the reason if needed. But, I think, it's better that the option is moved within `exec:` section (or system config) as written in TODO comment.

**Comment by muga** [comment]
Merging cc: @dmikurube

**Pull request merged**